src/compiler/crystal/tools/formatter.cr

```diff
@@ -1692,11 +1694,78 @@ module Crystal
 
       format_def_args node
 
-      format_macro_body node
+      if all_macro_literals?(node.body)
+        format_all_macro_literals(node, macro_node_line)
+      else
+        format_macro_body node
+      end
 
       false
     end
 
+    private def all_macro_literals?(node)
+      case node
+      when MacroLiteral
+        true
+      when Expressions
+        node.expressions.all? { |subnode| all_macro_literals?(subnode) }
+      else
+        false
+      end
+    end
+
+    private def gather_macro_literal_values(node)
+      String.build do |io|
+        gather_macro_literal_values(node, io)
+      end
+    end
+
+    private def gather_macro_literal_values(node, io)
+      case node
+      when MacroLiteral
+        io << node.value
+      when Expressions
+        node.expressions.each { |subnode| gather_macro_literal_values(subnode, io) }
+      else
+        false
+      end
+    end
+
+    private def format_all_macro_literals(node, macro_node_line)
+      source = gather_macro_literal_values(node.body)
+      begin
+        # Only format the macro contents if it's valid Crystal code
+        Parser.new(source).parse
```

Comment:
Here any `\{{` or `\{%` `MacroLiteral`s from `source` will be unescaped. Is there a situation where a macro's body is valid Crystal code if it has the escaped forms, but becomes invalid when they are escaped? (This should only affect a very small number of macros though.)

Reply:
It seems that this might be already handled. I tried formatting this:

```crystal
macro foo
  puts(\{{@type}}         )
end
```

And it actually produced this:

```crystal
macro foo
  puts(\{{@type}})
end
```

There's a `write_macro_slashes` method. I didn't take a look at how it works (I might have written that but now I can't remember.)

I'd say, let's first find a case where it fails.

Reply:
On further thought, this is probably fine because it just means the macro contents are invalid after any interpolation. This is what I got:

```crystal
{\{{ 1 }}}   # a TupleLiteral wrapping an escaped MacroLiteral
# becomes:
{{{ 1 }}}    # a MacroExpression wrapping a TupleLiteral

{ \{{ 1 }} } # a TupleLiteral wrapping an escaped MacroLiteral
# becomes:
{ {{ 1 }} }  # a TupleLiteral wrapping a MacroExpression

{\{{ 1 }} }  # a TupleLiteral wrapping an escaped MacroLiteral
# becomes:
{{{ 1 }} }   # invalid code
```

Reply:
Hm, I'm not understanding. Is there something to fix? Based on the comment above, it looks so but I'm not sure 😅 

Even if the macro contents are invalid, I don't think we should format those to something else.

Reply:
I think we can leave this check as is.